### DISCUSSION TITLE
How does monitoring change when moving to containers and orchestrators?

### DISCUSSION THREAD

**Discussion by plainlystated** [BretFisher/ama#149] [Q&A]
Once you have a few nodes up and running, what tools do you like for monitoring the health of the system (eg issues with managers, resource starvation, etc)?  Do you think about monitoring differently for a docker-based system?

**Comment by BretFisher** [comment]
Wow, how did I miss this one from 8 months ago?

Monitoring is both the same as *any other distributed system* and a bit of additional things to monitor for containers specifically. Most modern monitoring tools now have a "solution" for containers.

Remember that a containerized app is just a normal process on the host OS, so any host monitoring tool will see it, and be able to monitor the host and its Docker-based apps like normal.

The difference is you'll likely want to know things about the container settings surrounding that app, like, is it part of a multi-replica service, did you set resource limits and/or reservations for that container, and more.

The biggest change I think that your monitoring solution needs to take into account is the dynamic nature of your infrastructure once things are more ephemeral. Once you have Swarm, you can reschedule apps to different servers, and easily kill and replace server nodes. Old school monitoring systems have a harder time with this... and were designed in a world where servers and their apps lived a long time, and you created manual config files for each host "agent". With modern tools compatible with Swarm and Kubernetes, we expect them to listen to orchestrator events in the various API's and update their config based on the dynamic changing systems, rather than relying on a static config.

One great place to look for these tools is the CNCF landscape system, which lists projects that say they are designed for cloud-native container-based systems:

https://landscape.cncf.io/landscape=monitoring
